aiohttp/zlib_utils.py

```diff
@@ -0,0 +1,144 @@
+import asyncio
+import zlib
+from concurrent.futures import Executor
+from typing import Optional
+
+from typing_extensions import Protocol
+
+MAX_SYNC_CHUNK_SIZE = 1024
+
+
+class Compressor(Protocol):
+    def compress(self, data: bytes) -> bytes:
+        ...
+
+    def flush(self, mode: int = ...) -> bytes:
+        ...
+
+    def copy(self) -> "Compressor":
+        ...
+
+
+class Decompressor(Protocol):
```

Comment:
Given that the other use case is the custom BrotliDecoder class, why don't we just make both classes inherit from this, rather than using a Protocol?

Also, this class seems to be declared twice. It should probably just be declared in typedefs.py or similar.

Reply:
> Also, this class seems to be declared twice. It should probably just be declared in typedefs.py or similar.

The other protocol is different even though it has the same name. Also, they are not supposed to be public and are suited for internal usage in classes where they are declared on the module level. Having a declared protocol is anyway better that just Any type.

> Given that the other use case is the custom BrotliDecoder class, why don't we just make both classes inherit from this, rather than using a Protocol?

It is probably just a matter of taste. I like protocols because they just help to understand that a particular instance implements given methods or properties without applying too much of restrictions.

Reply:
> The other protocol is different even though it has the same name.

OK, I see these ones are just used to annotate the internal compressobj(). But, I don't see the point of doing that, mypy should already know the type of those objects, meaning that these 2 classes can be removed.

> I like protocols because they just help to understand that a particular instance implements given methods or properties without applying too much of restrictions.

Protocols are useful when you don't control the classes, or don't know the classes ahead of time. But, these protocols are being used where we are assigning hardcoded internal classes, so inheritance makes more sense.

Reply:
makes sense, removed